### ISSUE TITLE
Enforce strict boolean checks

### ISSUE THREAD

**Issue by ValentinCalomme** [python/mypy#16734]
**Feature**

A rule enforcing the strict use of boolean expressions in an if statement. It should be added to the "Miscellaneous strictness flags".

```
strict_conditional

    TYPE:
        boolean

    DEFAULT:
        False

    Enables or disables strict conditional checks. If True, `mypy` only allows boolean expressions in conditional expressions.
```

**Pitch**

I've encountered many hard-to-trace bugs due to non-boolean expressions used in conditionals. For instance, lists, containers, objects, or some literals being checked directly.

```py
if [1,2,3]: # list is not a boolean value, will be treated as False if list is empty or None
    ...

if "Hello world": # str is not a boolean value, will be treated as False if string is empty or None
    ...

if None: # not a boolean, will be treated as False if None
    ...
    
if 2: # not a boolean, might be treated as false if it equals 0
    ...
```

Very often, the developer's intention was to perform a null-check, or sometimes, an emptiness check, but this syntax yields potential edge cases where the developer did not intend the conditional to yield False.

With `strict_conditional` enabled, all of the above scenarios should be flagged.

The below examples naturally should always be valid as the expression yields a boolean value.

```py
if x < 0:
    ...

if isinstance(x, int):
    ...

if True:
    ...

if [1,2,3] is None:
    ....

if len({}) == 0:
    ....
```

This enforces explicit coding practices and better shows the intent of the developer.

**Comment by JamesParrott** [comment]
It's not a bad idea to have available in the ecosystem in general.  I just question if mypy (or any static type checker) is the best place to enforce programming style rules?

Don't any of the other established tools: e.g. Ruff, Black, Pylint, Pycodestyle (formerly pep8) already have a rule for this (or a similar rules in line with [PEP8](https://peps.python.org/pep-0008/#programming-recommendations) / or [Google's Python style guide](https://google.github.io/styleguide/pyguide.html#2.14) )?

**Comment by ValentinCalomme** [comment]
I created a similar issue for `ruff` (https://github.com/astral-sh/ruff/issues/9148), but they replied that it's out of the current scope as it would require type checking.

> I just question if mypy (or any static type checker) is the best place to enforce programming style rules.

I would argue that `mypy` already does this, albeit indirectly. Many flags and options aim to reduce errors created by type mismatches, which lead to enforcing some stylistic rules (i.e., specifically marking as optional due to the `strict_optional` rule.)

This rule does two things essentially:
- Make sure that null checks are explicit
- Reduce some bugs by making some checks explicit for containers (i.e. checking for an empty list)

The second point is indeed more of a bugbear type of situation, but the first could be seen as some sort of strictness flag related to None values.

**Comment by JelleZijlstra** [comment]
This isn't practical to do outside a type checker, as correctly noted on the Ruff issue. So this is indeed the correct place to request this feature.

Mypy already does a subset of what you ask for by e.g. warning when using a function in a boolean context (https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-function-isn-t-used-in-boolean-context-truthy-function). Warning for *all* non-booleans in a boolean context is too disruptive and too much against standard Python conventions to turn on by default or even in `--strict` mode, but it could be implemented as an opt-in error code.

**Comment by ValentinCalomme** [comment]
Indeed, I don't think it should be enabled by default.

**Comment by hauntsaninja** [comment]
(renamed the issue to avoid confusion with https://github.com/python/mypy/issues/8363)

**Comment by ValentinCalomme** [comment]
Oh, cool, interesting to see the discussion in that particular issue.

**Comment by FingalP** [comment]
+1 for this approach
@valentincalomme have you made any progress on this in the last couple of years?

**Comment by rambip** [comment]
The most complete discussion of this have been on the pyright repo: https://github.com/microsoft/pyright/discussions/6488

I can't believe we don't have this in 2026, it would be so valuable to check AI generated code !

**Comment by JamesParrott** [comment]
I've gained more sympathy for this as time's gone on (trying not to be a PEP 8 Hobgoblin, and embracing "explicit is better than implicit").  But firstly it's still an intended feature of Python, and not really a type bug.  Secondly, it's not even a cut-and-dried issue in the wider community.  

Pylint even encourages the opposing stance in [C1802](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/use-implicit-booleaness-not-len.html) and [C1803](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/use-implicit-booleaness-not-comparison.html), so lots of fun will be created for anyone using both Pylint and a type checker that implements this.
